### PR TITLE
feat: add Nat.powMod with GMP-backed extern for fast modular exponentiation

### PR DESCRIPTION
**Pull request by kim-em**
This PR adds `Nat.powMod b e m`, a modular exponentiation function provably equal to `b ^ e % m`, so large powers modulo a nonzero modulus can be evaluated without constructing the full power. It also makes exponentiation on `Fin n` use this operation.

The kernel-visible definition uses well-founded square-and-multiply recursion, allowing `decide` to reduce concrete modular powers in logarithmically many recursive calls. Compiled evaluation uses `mpz_powm` in GMP builds and an arbitrary-precision square-and-multiply fallback otherwise. When `m = 0`, the runtime delegates to `Nat.pow` to preserve Lean's `% 0` semantics; `Nat.pow` is adjusted so bases `0` and `1` work even when the exponent exceeds `UINT_MAX`, while other such powers retain the existing panic. The theorem `Nat.powMod_def` provides the mathematical characterization for symbolic reasoning and is deliberately not a simp lemma because eager expansion can turn cheap `powMod` goals into intractable powers.

`Fin.npow` and its `NatPow` instance move into `Init.Data.Fin.Lemmas` and use `Nat.powMod`, with corresponding `Fin.val_pow`, `Fin.pow_zero`, and `Fin.pow_succ` theorems. The new `Fin.reducePow` and `Nat.reducePowMod` simprocs let `simp` evaluate closed `Fin` powers and `Nat.powMod` terms without unfolding to `b ^ e % m`. Tests cover edge cases, agreement between kernel and native evaluation, very large concrete exponents, and the motivating 1024-bit `Fin` exponentiation case.

🤖 Prepared with Claude Code

### CODE REVIEW

File: src/runtime/mpz.cpp [hunk 1]
```diff
@@ -827,6 +834,25 @@ mpz mpz::pow(unsigned int p) const {
     return result;
 }
 
+mpz mpz::powm(mpz const & exp, mpz const & m) const {
+    lean_assert(!m.is_zero());
```
Comment (eric-wieser) [inline src/runtime/mpz.cpp:842]: Would a regular assert be better? What does gmp do?

File: src/Init/Data/Nat/PowMod.lean [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+the (potentially enormous) intermediate value `b ^ e`.
+
+Because `Nat.mod` satisfies `n % 0 = n`, we also have `powMod b e 0 = b ^ e`.
+
+The definition provided here is the logical model.
+-/
+@[expose, extern "lean_nat_powmod"]
+def powMod (b e m : @& Nat) : Nat := b ^ e % m
+
+theorem powMod_def (b e m : Nat) : powMod b e m = b ^ e % m := (rfl)
```
Comment (TwoFX) [inline src/Init/Data/Nat/PowMod.lean]: Why not make this a `simp` lemma and remove the rest of the API?

File: src/Init/Data/Nat/PowMod.lean [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+@[simp] theorem powMod_zero (b m : Nat) : powMod b 0 m = 1 % m := by
+  simp [powMod_def]
+
+@[simp] theorem powMod_one (b m : Nat) : powMod b 1 m = b % m := by
+  simp [powMod_def]
+
+theorem powMod_succ (b e m : Nat) : powMod b (e + 1) m = powMod b e m * b % m := by
```
Comment (TwoFX) [inline src/Init/Data/Nat/PowMod.lean]: This tripped me up when I first saw it, I feel it would be clearer as

```suggestion
theorem powMod_succ (b e m : Nat) : powMod b (e + 1) m = (powMod b e m * b) % m := by
```

File: src/runtime/object.cpp [hunk 4]
```diff
@@ -1558,6 +1558,17 @@ extern "C" LEAN_EXPORT lean_obj_res lean_nat_pow(b_lean_obj_arg a1, b_lean_obj_a
         return mpz_to_nat(mpz_value(a1).pow(lean_unbox(a2)));
 }
 
+extern "C" LEAN_EXPORT lean_obj_res lean_nat_powmod(b_lean_obj_arg b, b_lean_obj_arg e, b_lean_obj_arg m) {
+    mpz mb = lean_is_scalar(b) ? mpz::of_size_t(lean_unbox(b)) : mpz_value(b);
+    mpz me = lean_is_scalar(e) ? mpz::of_size_t(lean_unbox(e)) : mpz_value(e);
+    if (lean_is_scalar(m) && lean_unbox(m) == 0) {
+        // Lean convention: `_ % 0 = _`, so `Nat.powMod b e 0 = b ^ e`.
```
Comment (Rob23oba) [inline src/runtime/object.cpp:1636]: Why not just `return lean_nat_pow(b, e)`? `2 ^ (2 ^ 32)` is not supported, so it would probably be reasonable to also not support `(2).powMod (2 ^ 32) 0`. Also maybe then move the `mb` and `me` calculations after this check. Edit: if we want to support 1^huge and 0^huge, we should do so in `lean_nat_pow`.

File: src/Init/Data/Nat/PowMod.lean [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+runtime to use an efficient implementation (GMP's `mpz_powm`) which performs
+modular reduction at each step of repeated squaring. This avoids materializing
+the (potentially enormous) intermediate value `b ^ e`.
+
+Because `Nat.mod` satisfies `n % 0 = n`, we also have `powMod b e 0 = b ^ e`.
+
+The definition provided here is the logical model.
+-/
+@[expose, extern "lean_nat_powmod"]
+def powMod (b e m : @& Nat) : Nat := b ^ e % m
```
Comment (TwoFX) [inline src/Init/Data/Nat/PowMod.lean]: Have you considered doing a square-and-multiply implementation using `Nat.fix` here to also have good performance in the kernel?

File: src/Init/GrindInstances/Ring/Fin.lean [hunk 6]
```diff
@@ -13,26 +13,39 @@ public import Init.Data.Fin.Lemmas
...
 
 instance [NeZero n] : HPow (Fin n) Nat (Fin n) where
   hPow := Fin.npow
 
 instance [NeZero n] : Pow (Fin n) Nat where
   pow := Fin.npow
 
-@[simp] theorem pow_zero [NeZero n] (a : Fin n) : a ^ 0 = 1 := rfl
-@[simp] theorem pow_succ [NeZero n] (a : Fin n) (n : Nat) : a ^ (n+1) = a ^ n * a := rfl
+@[simp] theorem pow_zero [NeZero n] (a : Fin n) : a ^ 0 = 1 := by
```
Comment (TwoFX) [inline src/Init/GrindInstances/Ring/Fin.lean]: I guess it doesn't matter too much for low-level stuff like this, but the ratio of code smells to lines of code for this proof is very high.

Comment (kim-em) [reply src/Init/GrindInstances/Ring/Fin.lean]: I've cleaned up the proof.

File: src/runtime/mpz.cpp (hunk 1 above)
Comment (kim-em) [reply src/runtime/mpz.cpp:842]: 3dcca47bb3

File: src/Init/Data/Nat/PowMod.lean (hunk 2 above)
Comment (kim-em) [reply src/Init/Data/Nat/PowMod.lean]: Done in 31d1d03598. `powMod_def` is now `@[simp]`, so symbolic reasoning unfolds `powMod` to `b ^ e % m`.

Dropped the unused `powMod_one` / `powMod_mod` / `powMod_zero_right`; kept `powMod_zero` / `powMod_succ` (no longer `@[simp]`) as the named recurrence the `Fin` power instance uses.

File: src/Init/Data/Nat/PowMod.lean (hunk 5 above)
Comment (kim-em) [reply src/Init/Data/Nat/PowMod.lean]: Done in 38bb5a236a. `powMod`'s logical model is now square-and-multiply, so `decide` reduces it in `O(log e)` steps — a 1024-bit exponent in ~0.3s, where `b ^ e % m` couldn't be `decide`d at all.

I used a fuel parameter with *structural* recursion rather than `Nat.fix`.

Tests in 82d32edb7f. (Depth is inherently `log₂ e`, so cryptographic-scale exponents need `maxRecDepth` raised. Exponents up to ~2⁴⁸ are fine at the default.)

File: src/Init/Data/Nat/PowMod.lean [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+`b ^ e % m` provided `e < 2 ^ fuel`, reducing modulo `m` at each squaring step so
+that no intermediate exceeds `m * m`.
+
+The recursion is *structural* in `fuel` (so it reduces in the kernel, unlike a
+well-founded definition), while `e` is halved each step, so `powMod` — taking
+`fuel := e` — evaluates in `O(log e)` steps under `decide`. The current model
+`b ^ e % m` cannot: it would materialize the (astronomically large) `b ^ e`.
+-/
+@[expose]
+def powModAux (fuel b e m : Nat) : Nat :=
```
Comment (Rob23oba) [inline src/Init/Data/Nat/PowMod.lean]: Let's not hack together a fuel-based solution manually; `WellFounded.Nat.fix` is already fuel-based and will reduce well.

File: src/Init/GrindInstances/Ring/Fin.lean [hunk 8]
```diff
@@ -13,26 +13,33 @@ public import Init.Data.Fin.Lemmas
...
 namespace Fin
 
--- TODO: we should replace this at runtime with either repeated squaring,
--- or a GMP accelerated function.
 @[expose]
-def npow [NeZero n] (x : Fin n) (y : Nat) : Fin n := npowRec y x
+def npow [NeZero n] (x : Fin n) (y : Nat) : Fin n :=
+  ⟨Nat.powMod x.val y n, by rw [Nat.powMod_def]; exact Nat.mod_lt _ (Nat.pos_of_neZero n)⟩
+
+@[simp] theorem val_npow [NeZero n] (a : Fin n) (k : Nat) :
```
Comment (Rob23oba) [inline src/Init/GrindInstances/Ring/Fin.lean]: This is a bad simp lemma, one should write `a ^ k` and not `a.npow k`; a better lemma would be `val_pow` about `(a ^ k).val`; but perhaps this should also all be in a different place, i.e. `Init.Data.Fin.Basic` and `Init.Data.Fin.Lemmas`, since I wouldn't primarily associate `Fin` exponentiation with `grind`...

File: tests/elab/lean_nat_powmod.lean [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+/-! `e = 0` is `1 % m`. -/
+example : Nat.powMod 100 0 7 = 1 := rfl
+example : Nat.powMod 100 0 1 = 0 := rfl
+
+abbrev hugeExp : Nat := 4294967296
+
+-- Regression test: the runtime override must stay total for `m = 0`,
+-- even when the exponent exceeds `UINT_MAX`.
+#eval Nat.powMod 1 hugeExp 0 = 1
```
Comment (Rob23oba) [inline tests/elab/lean_nat_powmod.lean]: Since the only point of these tests seems to be to check that the result is true, use `#guard` instead.

File: tests/elab/lean_nat_powmod.lean [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+example : Nat.powMod 3 (2 ^ 40) 1000003 = 378344 := by native_decide
+
+/-!
+Large modular exponentiation (1024-bit prime), motivated by Mathlib's ZMod test.
+Prior to adding `lean_nat_powmod`, `Nat.powMod` would unfold to `b ^ e % m`
+which fails to terminate for this size of exponent.
+-/
+open Lean Elab Command in
+elab "#fast" c:command : command => do
+  let start ← IO.monoMsNow
```
Comment (Rob23oba) [inline tests/elab/lean_nat_powmod.lean]: Time is not a very reliable measure of speed; instead, try `set_option maxHeartbeats` with maybe twice the value it actually takes.

File: tests/elab/lean_nat_powmod.lean [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+abbrev M : Nat := 0xb10b8f96a080e01dde92de5eae5d54ec52c99fbcfb06a3c69a6a9dca52d23b616073e28675a23d189838ef1e2ee652c013ecb4aea906112324975c3cd49b83bfaccbdd7d90c4bd7098488e9c219a73724effd6fae5644738faa31a4ff55bccc0a151af5f0dc8b4bd45bf37df365c1a65e68cfda76d4da708df1fb2bc2e4a4371
+
+abbrev g : Nat := 0xa4d1cbd5c3fd34126765a442efb99905f8104dd258ac507fd6406cff14266d31266fea1e5c41564b777e690f5504f213160217b4b01b886a5e91547f9e2749f4d7fbd7d3b9a92ee1909d0d2263f80a76a6a24c087a091f531dbf0a0169b6a28ad662a4d18e73afa32d779d5918d08bc8858f4dcef97c2a24855e6eeb22b3b2e5
+
+-- Fermat's little theorem: if M is prime, `g^(M-1) ≡ 1 (mod M)`.
+#fast #eval Nat.powMod g (M - 1) M = 1
+
+-- And via `Fin`, which is the main motivating use case.
+#fast #eval ((Fin.ofNat _ g : Fin M) ^ (M - 1) = 1 : Bool)
+example : ((Fin.ofNat _ g : Fin M) ^ (M - 1)) = 1 := by native_decide
```
Comment (Rob23oba) [inline tests/elab/lean_nat_powmod.lean]: Do `#guard` here as above; also the `example` is kinda unnecessary then since it would check the same thing.

File: tests/elab/lean_nat_powmod.lean.out.expected [hunk 12]
```diff
@@ -0,0 +1,4 @@
+true
+true
+true
+true
```
Comment (Rob23oba) [inline tests/elab/lean_nat_powmod.lean.out.expected]: Just to make sure, you shouldn't need this file after replacing all `#eval`s with `#guard`s.

File: tests/elab/lean_nat_powmod.lean [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+Cross-check the compiled extern against the kernel's unfolding for
+inputs that exercise bignum arithmetic on both sides. A divergence here
+would make `native_decide` unsound.
+-/
+-- Fermat's little theorem on a small prime: `3^100 ≡ 1 (mod 101)`.
+example : Nat.powMod 3 100 101 = 1 := by native_decide
+example : Nat.powMod 3 100 101 = 1 := by decide
+-- Bignum base and modulus with a small exponent, so `decide` stays
+-- within the default exponentiation threshold.
+set_option exponentiation.threshold 400 in
```
Comment (Rob23oba) [inline tests/elab/lean_nat_powmod.lean]: This shouldn't be necessary with the new definition (remove the comment and `set_option`s if possible, use a bigger exponent and a smaller base).

File: tests/elab/lean_nat_powmod.lean [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+example : Nat.powMod 2 10 3 = 1  := rfl
+example : Nat.powMod 3 4 5   = 1  := rfl
+example : Nat.powMod 7 5 13  = 11 := rfl
+
+/-!
+The kernel can only evaluate `Nat.powMod` for small inputs (it unfolds to
+`b ^ e % m` and relies on the kernel's bignum override of `Nat.pow`).
+Larger inputs are tested via `#eval` below, which uses the `lean_nat_powmod`
+extern.
+-/
```
Comment (Rob23oba) [inline tests/elab/lean_nat_powmod.lean:6]: This is no longer correct and besides doesn't reflect what is tested below. Also, the basic tests already include edge cases, move the others there and maybe put a small comment on the right, e.g.
```lean-4
example : Nat.powMod 3 4 0  = 81 := rfl -- `powMod` behaves like `^` for modulus 0
```

File: tests/elab/lean_nat_powmod.lean [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+example : Nat.powMod 100 0 1 = 0 := rfl
+
+abbrev hugeExp : Nat := 4294967296
+
+-- Regression test: the runtime override must stay total for `m = 0`,
+-- even when the exponent exceeds `UINT_MAX`.
+#eval Nat.powMod 1 hugeExp 0 = 1
+#eval Nat.powMod 0 hugeExp 0 = 0
+example : Nat.powMod 1 hugeExp 0 = 1 := by native_decide
+example : Nat.powMod 0 hugeExp 0 = 0 := by native_decide
```
Comment (Rob23oba) [inline tests/elab/lean_nat_powmod.lean]: These are duplicates of the `#eval`s (to become `#guard`s) above.

File: tests/elab/lean_nat_powmod.lean [hunk 16]
```diff
@@ -0,0 +1,102 @@
+import Lean
```
Comment (Rob23oba) [inline tests/elab/lean_nat_powmod.lean]: If you just use a `maxHeartbeats` test, you shouldn't need the `Lean` import.

File: tests/elab/lean_nat_powmod.lean [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+example : Nat.powMod 3 4 0  = 81   := rfl
+
+/-! `b = 0` with positive exponent is `0 % m`. -/
+example : Nat.powMod 0 5 7 = 0 := rfl
+
+/-! `e = 0` is `1 % m`. -/
+example : Nat.powMod 100 0 7 = 1 := rfl
+example : Nat.powMod 100 0 1 = 0 := rfl
+
+abbrev hugeExp : Nat := 4294967296
```
Comment (Rob23oba) [inline tests/elab/lean_nat_powmod.lean]: 4294967296 maybe looks too much like a random number, write `2 ^ 32` instead.

File: src/Init/GrindInstances/Ring/Fin.lean [hunk 18]
```diff
@@ -158,6 +165,9 @@ instance [i : NeZero n] : ToInt.Pow (Fin n) (.co 0 n) where
 
 instance : PowIdentity (Fin 2) 2 where
   pow_eq x := by
+    ext
+    show (Fin.npow x 2).val = x.val
+    rw [val_npow]
```
Comment (Rob23oba) [inline src/Init/GrindInstances/Ring/Fin.lean]: If you instead put a `Fin.val_pow` lemma in `Init.Data.Fin.Lemmas` as suggested above, you shouldn't need the `show`.

**Review by Rob23oba — CHANGES_REQUESTED** [review]

File: src/Init/Data/Nat/PowMod.lean (hunk 5 above)
Comment (kim-em) [reply src/Init/Data/Nat/PowMod.lean]: ed87f476cc

File: src/Init/Data/Nat/PowMod.lean (hunk 7 above)
Comment (kim-em) [reply src/Init/Data/Nat/PowMod.lean]: ed87f476cc

File: src/Init/GrindInstances/Ring/Fin.lean (hunk 18 above)
Comment (kim-em) [reply src/Init/GrindInstances/Ring/Fin.lean]: 3573b3dfd5

File: tests/elab/lean_nat_powmod.lean (hunk 9 above)
Comment (kim-em) [reply tests/elab/lean_nat_powmod.lean]: 3351fee2bb

File: tests/elab/lean_nat_powmod.lean (hunk 15 above)
Comment (kim-em) [reply tests/elab/lean_nat_powmod.lean]: 3351fee2bb

File: tests/elab/lean_nat_powmod.lean (hunk 11 above)
Comment (kim-em) [reply tests/elab/lean_nat_powmod.lean]: 3351fee2bb

File: tests/elab/lean_nat_powmod.lean (hunk 10 above)
Comment (kim-em) [reply tests/elab/lean_nat_powmod.lean]: 3351fee2bb

File: tests/elab/lean_nat_powmod.lean (hunk 13 above)
Comment (kim-em) [reply tests/elab/lean_nat_powmod.lean]: 3351fee2bb

File: tests/elab/lean_nat_powmod.lean (hunk 14 above)
Comment (kim-em) [reply tests/elab/lean_nat_powmod.lean:6]: 3351fee2bb

File: tests/elab/lean_nat_powmod.lean (hunk 17 above)
Comment (kim-em) [reply tests/elab/lean_nat_powmod.lean]: 3351fee2bb

File: tests/elab/lean_nat_powmod.lean (hunk 16 above)
Comment (kim-em) [reply tests/elab/lean_nat_powmod.lean]: 3351fee2bb

File: tests/elab/lean_nat_powmod.lean.out.expected (hunk 12 above)
Comment (kim-em) [reply tests/elab/lean_nat_powmod.lean.out.expected]: 3351fee2bb

File: src/Init/GrindInstances/Ring/Fin.lean (hunk 8 above)
Comment (kim-em) [reply src/Init/GrindInstances/Ring/Fin.lean]: 3573b3dfd5 (oops, not quite) and then 58b03306574b82545f5ce99bda4256d6ec7fb0a9.

File: src/Init/Data/Fin/Lemmas.lean [hunk 19]
```diff
@@ -1217,4 +1218,27 @@ protected theorem mul_zero [NeZero n] (k : Fin n) : k * 0 = 0 := by
...
+extern. `npow x y` computes `x ^ y` with the result reduced modulo `n`.
+-/
+def npow [NeZero n] (x : Fin n) (y : Nat) : Fin n :=
+  ⟨Nat.powMod x.val y n, by rw [Nat.powMod_def]; exact Nat.mod_lt _ (Nat.pos_of_neZero n)⟩
+
+instance [NeZero n] : HPow (Fin n) Nat (Fin n) where
+  hPow := Fin.npow
+
+instance [NeZero n] : Pow (Fin n) Nat where
+  pow := Fin.npow
```
Comment (Rob23oba) [inline src/Init/Data/Fin/Lemmas.lean:1229]: I would've moved this to the `Init.Data.Fin.Basic` file since there are only very few definitions in `Lemmas` and especially none of the arithmetic (unary and binary) operations. So I would put the `npow` definition after `div` and the instance (only a `NatPow` instance) after the `instance : Div`.

File: src/Init/Data/Fin/Lemmas.lean [hunk 20]
```diff
@@ -1217,4 +1218,27 @@ protected theorem mul_zero [NeZero n] (k : Fin n) : k * 0 = 0 := by
 protected theorem zero_mul [NeZero n] (k : Fin n) : (0 : Fin n) * k = 0 := by
   simp [Fin.ext_iff, mul_def]
 
+/--
+Exponentiation on `Fin n` by repeated squaring, backed by the fast `Nat.powMod`
+extern. `npow x y` computes `x ^ y` with the result reduced modulo `n`.
+-/
+def npow [NeZero n] (x : Fin n) (y : Nat) : Fin n :=
+  ⟨Nat.powMod x.val y n, by rw [Nat.powMod_def]; exact Nat.mod_lt _ (Nat.pos_of_neZero n)⟩
```
Comment (Rob23oba) [inline src/Init/Data/Fin/Lemmas.lean]: This (and similarly the instance) don't require `NeZero`, we have `x.pos`.

File: src/Init/Data/Nat/PowMod.lean [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+
+The logical model is square-and-multiply, halving `e` at each step and reducing
+modulo `m` so that no intermediate exceeds `m * m`. It is defined by well-founded
+recursion via `WellFounded.Nat.fix`, which reduces in the kernel, so `powMod`
+evaluates in `O(log e)` steps under `decide` and `rfl`. The naive model
+`b ^ e % m` cannot: it would materialize the (astronomically large) `b ^ e`.
+
+The `@[semireducible]` attribute keeps the definition unfoldable by `decide` and
+`rfl`; without it, well-founded definitions are marked `@[irreducible]`, which
+blocks reduction at default transparency before the kernel is ever reached.
```
Comment (Rob23oba) [inline src/Init/Data/Nat/PowMod.lean]: Please follow the documentation style guidelines for constants here (see [here](https://github.com/leanprover/lean4/blob/7c8494ef0b9e5e72681449535f9ccdf5f543b4f5/doc/style.md#constants)). But outside of what's mentioned in the style guidelines, I would not mention implementation details here, notably:
- The specific GMP function and the algorithm that function supposedly uses
- The fact that it uses `semireducible` and wellfounded recursion (at most, say something like "instead of the naive definition `b ^ e % m`, the logical model uses square-and-multiply to keep reduction fast for large exponents")

These facts are not useful for any user of the function, the only things you really need to know are
- This is basically `b ^ e % m` but fast
- But there's a reason why it's not defeq to `b ^ e % m`
- How it behaves in edge cases

File: src/Init/Data/Fin/Lemmas.lean [hunk 22]
```diff
@@ -1217,4 +1218,27 @@ protected theorem mul_zero [NeZero n] (k : Fin n) : k * 0 = 0 := by
...
+  pow := Fin.npow
+
+@[simp] theorem val_pow [NeZero n] (a : Fin n) (k : Nat) :
+    (a ^ k).val = a.val ^ k % n :=
+  Nat.powMod_def a.val k n
+
+@[simp] theorem pow_zero [NeZero n] (a : Fin n) : a ^ 0 = 1 := by
+  ext; exact Nat.powMod_zero a.val n
+
+@[simp] theorem pow_succ [NeZero n] (a : Fin n) (k : Nat) : a ^ (k + 1) = a ^ k * a := by
```
Comment (Rob23oba) [inline src/Init/Data/Fin/Lemmas.lean]: Don't make this a `simp` lemma, we also don't have `pow_succ` as simp in mathlib because it would expand `a ^ 34` into repeated multiplication.

File: src/Init/GrindInstances/Ring/Fin.lean [hunk 23]
```diff
@@ -122,8 +108,8 @@ instance (n : Nat) [NeZero n] : CommRing (Fin n) where
   mul_one := Fin.mul_one
   left_distrib := Fin.left_distrib
   zero_mul := Fin.zero_mul
-  pow_zero _ := by rfl
-  pow_succ _ _ := by rfl
+  pow_zero a := Fin.pow_zero a
+  pow_succ a k := Fin.pow_succ a k
```
Comment (Rob23oba) [inline src/Init/GrindInstances/Ring/Fin.lean]:
```suggestion
  pow_zero := Fin.pow_zero
  pow_succ := Fin.pow_succ
```

File: src/Init/Data/Nat/PowMod.lean [hunk 24]
```diff
@@ -0,0 +1,73 @@
...
+`powMod` agrees with the naive `b ^ e % m`. Marked `@[simp]` so that symbolic
+reasoning unfolds to the naive form (with its full lemma library); concrete
+evaluation by `decide` instead reduces the efficient definition and is unaffected
+by this lemma.
+-/
+@[simp] theorem powMod_def (b e m : Nat) : powMod b e m = b ^ e % m := by
+  fun_induction powMod b e m with
+  | case1 b => rw [Nat.pow_zero]
+  | case2 b e hne r hodd ih =>
+    show powMod (b * b % m) (e / 2) m * b % m = b ^ e % m
```
Comment (Rob23oba) [inline src/Init/Data/Nat/PowMod.lean]:
```suggestion
    subst r
```

File: src/Init/Data/Nat/PowMod.lean [hunk 25]
```diff
@@ -0,0 +1,73 @@
...
+    have hod : 2 * (e / 2) + 1 = e := by omega
+    rw [ih, ← Nat.pow_mod, ← Nat.pow_two, ← Nat.pow_mul, Nat.mod_mul_mod,
+      ← Nat.pow_succ, Nat.succ_eq_add_one, hod]
+  | case3 b e hne r heven ih =>
+    show powMod (b * b % m) (e / 2) m = b ^ e % m
```
Comment (Rob23oba) [inline src/Init/Data/Nat/PowMod.lean]:
```suggestion
    subst r
```

File: src/Init/Data/Fin/Lemmas.lean [hunk 26]
```diff
@@ -1217,4 +1218,27 @@ protected theorem mul_zero [NeZero n] (k : Fin n) : k * 0 = 0 := by
...
+  hPow := Fin.npow
+
+instance [NeZero n] : Pow (Fin n) Nat where
+  pow := Fin.npow
+
+@[simp] theorem val_pow [NeZero n] (a : Fin n) (k : Nat) :
+    (a ^ k).val = a.val ^ k % n :=
+  Nat.powMod_def a.val k n
+
+@[simp] theorem pow_zero [NeZero n] (a : Fin n) : a ^ 0 = 1 := by
```
Comment (Rob23oba) [inline src/Init/Data/Fin/Lemmas.lean]: This lemma (and the one after) should be `protected` since they also exist globally in mathlib.

File: src/Init/Data/Nat/PowMod.lean [hunk 27]
```diff
@@ -0,0 +1,73 @@
...
+    if e % 2 = 1 then r * b % m else r
+termination_by e
+decreasing_by omega
+
+/--
+`powMod` agrees with the naive `b ^ e % m`. Marked `@[simp]` so that symbolic
+reasoning unfolds to the naive form (with its full lemma library); concrete
+evaluation by `decide` instead reduces the efficient definition and is unaffected
+by this lemma.
+-/
```
Comment (Rob23oba) [inline src/Init/Data/Nat/PowMod.lean]: This docstring is unnecessary, the fact that this is a `simp` lemma tells you everything you need to know about it.

File: src/runtime/mpz.cpp [hunk 28]
```diff
@@ -832,6 +854,40 @@ mpz mpz::pow(unsigned int p) const {
...
+    mpz e(exp);
+    while (!e.is_zero()) {
+        if (e.mod8() & 1) {
+            result *= base;
+        }
+        base *= base;
+        div2k(e, e, 1);
+    }
+    return result;
+}
```
Comment (Rob23oba) [inline src/runtime/mpz.cpp]: This is unused since we abort early on big exponents already.

File: src/runtime/mpz.cpp [hunk 29]
```diff
@@ -214,6 +214,28 @@ mpz mpz::pow(unsigned int exp) const {
...
```
Comment (Rob23oba) [inline src/runtime/mpz.cpp]: This is unused since we abort early on big exponents already.

File: src/runtime/mpz.h [hunk 30]
```diff
@@ -210,6 +210,10 @@ class LEAN_EXPORT mpz {
     friend mpz rem(mpz const & a, mpz const & b) { mpz r(a); return r %= b; }
 
     mpz pow(unsigned int exp) const;
+    /** \brief Exponentiation with an arbitrary-size natural exponent. */
+    mpz pow(mpz const & exp) const;
```
Comment (Rob23oba) [inline src/runtime/mpz.h]: This is unused since we abort early on big exponents already.

**Review by Rob23oba — CHANGES_REQUESTED** [review]

File: src/Init/Data/Fin/Lemmas.lean (hunk 19 above)
Comment (kim-em) [reply src/Init/Data/Fin/Lemmas.lean:1229]: As far as I can see this causes an import cycle.

File: src/Init/Data/Nat/PowMod.lean (hunk 2 above)
Comment (kim-em) [reply src/Init/Data/Nat/PowMod.lean]: Hmm, actually I'm changing my mind here. I don't like that `simp` then evaluates ground terms with Nat.powMod. I think I'd prefer to leave this for an explicit unfolding.

File: src/Init/Data/Fin/Lemmas.lean (hunk 19 above)
Comment (Rob23oba) [reply src/Init/Data/Fin/Lemmas.lean:1229]: Alright; I see your concern; `Nat.pow_mod` is in `Init.Data.Nat.Lemmas` which imports `Init.Data.Fin.Basic`. In that case, it's probably fine as it is now.

File: src/Lean/Meta/Tactic/Simp/BuiltinSimprocs/Fin.lean [hunk 31]
```diff
@@ -79,6 +79,17 @@ set_option linter.coreInternal.internalModule false in -- User-facing builtin si
...
+set_option linter.coreInternal.internalModule false in -- User-facing builtin simprocs are fine
+builtin_dsimproc [simp, seval] reducePow ((_ ^ _ : Fin _)) := fun e => do
+  let_expr HPow.hPow _ _ _ inst a k := e | return .continue
+  let some v ← fromExpr? a | return .continue
+  let some k ← getNatValue? k | return .continue
+  -- A different `Pow (Fin n) Nat` instance need not agree with `Fin.npow` definitionally.
+  let finTy := mkApp (mkConst ``Fin) (toExpr v.n)
+  let instFin := mkApp3 (mkConst ``instHPow [.zero, .zero]) finTy (mkConst ``Nat)
+    (mkApp2 (mkConst ``instPowNat [.zero]) finTy (mkApp (mkConst ``Fin.instNatPow) (toExpr v.n)))
+  unless ← matchesInstance inst instFin do return .continue
```
Comment (Rob23oba) [inline src/Lean/Meta/Tactic/Simp/BuiltinSimprocs/Fin.lean]: Hmm, I'm not sure about this; I mean it *is* the correct thing to do here but at the same time it is something that no other simproc here does. I feel it's better to instead fix this properly for all at once and not do it here yet. For this purpose it'd be nice to add all the necessary instances to `Lean.Expr`, similar to `Nat.mkInstHAdd`, and use these in all the simprocs to check the instances.

**Review by Rob23oba — APPROVED** [review]
Alright thanks, I'm happy with this now!
